Sming/Arch/Esp32/Core/HardwarePWM.cpp

```diff
@@ -0,0 +1,305 @@
+ * ==================================================
+ * currently, the code just uses a % 8 operation on the pin index to calculate whether to assign a pin to either
+ * high speed or low speed pwm blocks. This doesn't make a whole lot of sense since it makes it impossible
+ * for Sming devs to actually use the functionality behind it. 
+ * Also, it currently does not reflect the fact that different SOCs have a different number of channels per block
+ * (specifically, the esp32c3 only has six channels and no highspeed mode). 
+ * I will continue in two ways: 
+ * - implement the "vanilla" Sming HardwarePWM interface that will hide the underlying architecture but allow up to 16 
+ *   channels on an ESP32 
+ * - implement overloads for the relevant functions that allow selecting hs mode where applicable. 
+ * 
+ * ToDo: implement PWM bit width control
+ * =====================================
+ * the current HardwarePWM implementation does not care about the PWM timer bit width. To leverage the functionality 
+ * of the ESP32 hardware, it's necessary to make this configurable. As the width is per timer and all the Sming defined
+ * functions are basically per pin, this needs to be an extension of the overal model, exposing at least timers. 
+ * This, too, will require a compatible "basic" interface and an advanced interface that allows assiging pins (channels) 
+ * to timers and the configuration of the timers themselves. 
+ * The esp_idf does not provide a way to read the bit width configured for a channel, but I think it'll be useful to be able
+ * to read back this value, not least to find the value for getMaxDuty() for a channel. It will either have to be stored in the
+ * module or maybe read from the hardware directly (LEDC_[HL]STIMERx_CONF_REG & 0x1f)
+ * 
+ * hardware technical reference: 
+ * =============================
+ * https://www.espressif.com/sites/default/files/documentation/esp32_technical_reference_manual_en.pdf#ledpwm
+ * 
+ * Overview of the whole ledc-system here: 
+ * https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/ledc.html
+ * 
+ * the ledc interface also exposes some advanced functions such as fades that are then done in hardware. 
+ * ToDo: implement a Sming interface for fades
+ * 
+ */
+
+#include <HardwarePWM.h>
+#include <debug_progmem.h>
+#include <driver/periph_ctrl.h>
+#include <driver/ledc.h>
+#include <esp_err.h>
+#include <hal/ledc_types.h>
+
+#define DEFAULT_RESOLUTION ledc_timer_bit_t(10)
+#define DEFAULT_PERIOD 200
+namespace
+{
+ledc_channel_t pinToChannel(uint8_t pin)
+{
+	return ledc_channel_t(pin % 8);
+}
+
+ledc_mode_t pinToGroup(uint8_t pin)
+{
+	return ledc_mode_t(pin / 8);
+}
+
+ledc_timer_t pinToTimer(uint8_t pin)
+{
+	return ledc_timer_t((pin / 2) % 4);
+}
+
+uint32_t periodToFrequency(uint32_t period)
+{
+	return (period == 0) ? -1 : (1000000 / period);
+}
+
+uint32_t frequencyToPeriod(uint32_t freq)
+{
+	return (freq == 0) ? -1 : (1000000 / freq);
+}
```

Comment:
Is it intentional to return `-1` for invalid values? Should it be `0`?

Reply:
you're right, -1 doesn't even make sense as it's uint32_t. I guess I was thinking that 0 could be an almost valid response but that isn't true either. I'll change it to 0